### ISSUE TITLE
Engine fails under load

### ISSUE THREAD

**Issue by lalexgap** [statechannels/go-nitro#744]
Running the testground integration test which is creating and closing virtual channels I often run into this error:

```
cannot transfer ownership of channel to from objective VirtualFund-0x86e2d9f554f8529230e8431b9757a1a0ce10417481d993c1741c3a82c1cbacda to VirtualDefund-0x86e2d9f554f8529230e8431b9757a1a0ce10417481d993c1741c3a82c1cbacda
```

**Comment by lalexgap** [comment]
> > Since only one objective can own a ledger at a time things blow up.
> 
> Do you mean, only one objective can own a _virtual_ channel at a time? As discussed at standup today, ledger channels are expected to "feature" in multiple objectives (for different virtual channels) concurrently.
> 

That's correct! I confused myself with ownership of virtual channels and ledger channels. The issue is that only one objective can own the **virtual channel**

**Comment by geoknee** [comment]
What's the correct long-term behaviour for this? How about we let the **defund** objective forcefully take over from the **fund** objective? Is there any point in the client in question proceeding to try and fund the channel when they know at least one peer is trying to defund it?

**Comment by lalexgap** [comment]
> Is there any point in the client in question proceeding to try and fund the channel when they know at least one peer is trying to defund it?

I don't think this scenario would occur. A peer can only start virtual defunding once the virtual funding objective is complete. So the peer would only start virtual defunding once they have received signed post fund setups from everyone to complete virtual funding.

What can happen now is that we happen to receive the virtual defund message **before** we receive their post fund setup from a peer. Since the virtual funding objective still has ownership on the virtual channel our store errors and the engine panics.

We only make [between 1 and 5 payments](https://github.com/statechannels/go-nitro-testground/blob/698ff71bd11b4bc352799aa83c53c354a5e56881/tests/virtual-payment.go#L120) before attempting to close the virtual channel so virtual channels are pretty "short-lived", which is why we run into this.

For now we avoid this by [sleeping](https://github.com/statechannels/go-nitro-testground/blob/698ff71bd11b4bc352799aa83c53c354a5e56881/tests/virtual-payment.go#L131) before closing a virtual channel. This ensures that our peers have a had a chance to receive our post fund setups and finish their virtual funding objectives.

> What's the correct long-term behaviour for this? 

In the short term it's probably good enough to catch this scenario and report an error. Right now the whole engine panics when this happens.

Longer term I think we'd want something like Option 1 we discussed [here](https://github.com/statechannels/go-nitro/issues/883). We should queue up the virtual defund objective and start cranking after the virtual fund objective is complete.

**Comment by geoknee** [comment]
Yeah I think sleeping is not the long term solution. I still think that abandoning the virtual fund objective might be the right thing to do in this case, if we think about the semantic meaning of each of those messages coming in. Shall we discuss at standup today?

**Comment by geoknee** [comment]
Discussed in standup today: abandoning the `virtualfund` objective is not a good approach, since validating a proposal to virtually defund requires inspecting **valid vouchers received**, which depends on having the postfund signatures (i.e. completing `virtualfund`). Therefore the best approach is to queue up the `virtualdefund` objective and make sure it starts executing once the `virtualfund` objective completes.
